solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/HeuristicBoostedModel.java

```diff
@@ -0,0 +1,177 @@
+    initHeuristicBoost();
+  }
+
+  protected void initHeuristicBoost() {
+    Map<String, Object> params = (Map<String, Object>) HeuristicBoostedModel.this.getParams().get(HEURISTIC_BOOST_PARAM);
+    if (params != null) {
+      this.heuristicFeatureBoost = new EuristicFeatureBoost(params);
+    }
+  }
+
+  @Override
+  public float score(float[] modelFeatureValuesNormalized) {
+    float internalModelScore = scoringAlgorithm.score(modelFeatureValuesNormalized);
+    float finalScore = internalModelScore;
+    if (heuristicFeatureBoost != null) {
+      float boostFeatureValue = modelFeatureValuesNormalized[heuristicFeatureBoost.featureIndex];
+      float weightedBoostFeatureValue = getWeightedBoostFeatureValue(internalModelScore, boostFeatureValue);
+
+      switch (heuristicFeatureBoost.type) {
+        case product:
+          finalScore *= weightedBoostFeatureValue;
+          break;
+
+        case sum:
+          finalScore += weightedBoostFeatureValue;
+          break;
+      }
+    }
+    return finalScore;
+  }
+
+  private float getWeightedBoostFeatureValue(float internalModelScore, float boostFeatureValue) {
+    float weightedBoostFeatureValue = heuristicFeatureBoost.weight * boostFeatureValue;
+    switch (heuristicFeatureBoost.type) {
+      case product:
+        if ((internalModelScore < 0 && weightedBoostFeatureValue > 0) || (internalModelScore > 0 && weightedBoostFeatureValue < 0)) {
+          weightedBoostFeatureValue = 1 / weightedBoostFeatureValue;
+        }
+        break;
+    }
+    return weightedBoostFeatureValue;
+  }
+
+  @Override
+  public Explanation explain(LeafReaderContext context, int doc, float finalScore, List<Explanation> featureExplanations) {
+    final float[] modelFeatureValuesNormalized = getFeatureVector(featureExplanations);
+    final float internalModelScore = scoringAlgorithm.score(modelFeatureValuesNormalized);
+    if (heuristicFeatureBoost == null) {
+      return scoringAlgorithm.explain(context, doc, internalModelScore, featureExplanations);
+    }
+
+    final List<Explanation> details = new ArrayList<>();
+    final float boostFeatureValue = modelFeatureValuesNormalized[heuristicFeatureBoost.featureIndex];
+    float weightedBoostFeatureValue = getWeightedBoostFeatureValue(internalModelScore, boostFeatureValue);
+    final Explanation boostExplain = Explanation.match(weightedBoostFeatureValue,
+        heuristicFeatureBoost.weight + " weight on feature [" + heuristicFeatureBoost.feature + "] : " + boostFeatureValue);
+    details.add(boostExplain);
+    final Explanation internalModelExplain = scoringAlgorithm.explain(context, doc, internalModelScore, featureExplanations);
+    details.add(internalModelExplain);
+
+    return Explanation.match(finalScore, toString()
+        + " model applied to features, " + heuristicFeatureBoost.type + " of:", details);
+  }
+
+  private float[] getFeatureVector(List<Explanation> featureExplanations) {
```

Comment:
This looks like the same thing LambdaMART does.  Move this up into a common util function for both to use.